stdlib/Test/src/Test.jl

```diff
@@ -1938,39 +1918,21 @@ end
 #-----------------------------------------------------------------------
 # Various helper methods for test sets
 
+const CURRENT_TESTSET = ScopedValue{AbstractTestSet}(FallbackTestSet())
+const TESTSET_DEPTH = ScopedValue{Int}(0)
+
+macro with_testset(ts, expr)
+    :(@with(CURRENT_TESTSET => $(esc(ts)), TESTSET_DEPTH => get_testset_depth() + 1, $(esc(expr))))
+end
+
 """
     get_testset()
 
 Retrieve the active test set from the task's local storage. If no
 test set is active, use the fallback default test set.
 """
 function get_testset()
-    testsets = get(task_local_storage(), :__BASETESTNEXT__, AbstractTestSet[])
-    return isempty(testsets) ? fallback_testset : testsets[end]
-end
-
-"""
-    push_testset(ts::AbstractTestSet)
-
-Adds the test set to the `task_local_storage`.
-"""
-function push_testset(ts::AbstractTestSet)
-    testsets = get(task_local_storage(), :__BASETESTNEXT__, AbstractTestSet[])
-    push!(testsets, ts)
-    setindex!(task_local_storage(), testsets, :__BASETESTNEXT__)
-end
```

Comment:
Apparently some (testing) packages were using this: https://juliahub.com/ui/Search?type=symbols&q=push_testset&u=use.  Presumably it's ok to ask them to adapt their use of internal functions?  [These aren't documented](https://docs.julialang.org/en/v1/stdlib/Test/).

Reply:
Yes, unfortunately I don't see how we could keep that working

Reply:
How would one update a codebase that is using `push_testset`? is it as simple as changing from
```julia
push_testset(ts)
try
   ...
finally
  pop_testset()
end
```
to
```julia
@with_testset ts begin
  ...
end
```
?

(p.s. Very pleased to see this improvement!)

Reply:
Yes, that should work I think.